File: src/db-toolkit/core/routes/data_explorer.py
"""Data explorer routes for browsing table data."""

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel
from typing import Optional
from core.storage import ConnectionStorage
from operations.data_explorer import DataExplorer

router = APIRouter()
storage = ConnectionStorage()
explorer = DataExplorer()


class DataBrowseRequest(BaseModel):
    """Data browse request."""
    schema_name: str
    table_name: str
    limit: int = 100
    offset: int = 0
    sort_column: Optional[str] = None
    sort_order: Optional[str] = "ASC"
    filters: Optional[dict] = None


@router.post("/connections/{connection_id}/data/browse")
async def browse_table_data(connection_id: str, request: DataBrowseRequest):
    """Browse table data with pagination, sorting, and filtering."""
    from operations.operation_lock import operation_lock

    connection = await storage.get_connection(connection_id)
    if not connection:
        raise HTTPException(status_code=404, detail="Connection not found")

    if operation_lock.is_locked(connection_id):
        raise HTTPException(
            status_code=409, detail="Connection is busy with another operation"
        )

    lock = operation_lock.get_lock(connection_id)
    async with lock:
        result = await explorer.browse_data(
            connection=connection,
            schema_name=request.schema_name,
            table_name=request.table_name,
            limit=request.limit,
            offset=request.offset,
            sort_column=request.sort_column,
            sort_order=request.sort_order,
            filters=request.filters,
        )
        return result


@router.get("/connections/{connection_id}/data/count")
async def get_table_row_count(
    connection_id: str,
    schema_name: str = Query(...),
    table_name: str = Query(...)
):
    """Get total row count for a table."""
    connection = await storage.get_connection(connection_id)
    if not connection:
        raise HTTPException(status_code=404, detail="Connection not found")

    count = await explorer.get_row_count(connection, schema_name, table_name)
    return {"success": True, "count": count}


@router.get("/connections/{connection_id}/data/relationships")
async def get_table_relationships(
    connection_id: str,
    schema_name: str = Query(...),
    table_name: str = Query(...)
):
    """Get foreign key relationships for a table."""
    connection = await storage.get_connection(connection_id)
    if not connection:
        raise HTTPException(status_code=404, detail="Connection not found")

    result = await explorer.get_table_relationships(connection, schema_name, table_name)
    return result


class CellDataRequest(BaseModel):
    """Cell data request."""
    schema_name: str
    table_name: str
    column_name: str
    row_identifier: dict


@router.post("/connections/{connection_id}/data/cell")
async def get_cell_data(connection_id: str, request: CellDataRequest):
    """Get full content of a specific cell."""
    connection = await storage.get_connection(connection_id)
    if not connection:
        raise HTTPException(status_code=404, detail="Connection not found")

    result = await explorer.get_cell_data(
        connection,
        request.schema_name,
        request.table_name,
        request.column_name,
        request.row_identifier
    )
    return result

File: src/db-toolkit/electron/backend/handlers/data-explorer.js
/**
 * Data explorer IPC handlers.
 */

const { ipcMain } = require('electron');
const { connectionStorage } = require('../utils/connection-storage');
const { connectionManager } = require('../utils/connection-manager');
const { dataExplorer } = require('../operations/data-explorer');
const { logger } = require('../utils/logger.js');

function registerDataExplorerHandlers() {
  // Browse table data
  ipcMain.handle('dataExplorer:browse', async (event, connectionId, request) => {
    try {
      const connection = await connectionStorage.getConnection(connectionId);
      if (!connection) {
        throw new Error('Connection not found');
      }

      const result = await dataExplorer.browseData(
        connection,
        request.schema_name,
        request.table_name,
        request.limit || 100,
        request.offset || 0,
        request.sort_column,
        request.sort_order || 'ASC',
        request.filters
      );
      return { data: result };
    } catch (error) {
      logger.error('Browse data error:', error);
      throw error;
    }
  });

  // Get row count
  ipcMain.handle('dataExplorer:count', async (event, connectionId, schemaName, tableName) => {
    try {
      const connection = await connectionStorage.getConnection(connectionId);
      if (!connection) {
        throw new Error('Connection not found');
      }

      const count = await dataExplorer.getRowCount(connection, schemaName, tableName);
      return { data: { success: true, count } };
    } catch (error) {
      logger.error('Get row count error:', error);
      throw error;
    }
  });

  // Get table relationships
  ipcMain.handle('dataExplorer:relationships', async (event, connectionId, schemaName, tableName) => {
    try {
      const connection = await connectionStorage.getConnection(connectionId);
      if (!connection) {
        throw new Error('Connection not found');
      }

      const result = await dataExplorer.getTableRelationships(connection, schemaName, tableName);
      return { data: result };
    } catch (error) {
      logger.error('Get relationships error:', error);
      throw error;
    }
  });

  // Get cell data
  ipcMain.handle('dataExplorer:cell', async (event, connectionId, request) => {
    try {
      const connection = await connectionStorage.getConnection(connectionId);
      if (!connection) {
        throw new Error('Connection not found');
      }

      const result = await dataExplorer.getCellData(
        connection,
        request.schema_name,
        request.table_name,
        request.column_name,
        request.row_identifier
      );
      return { data: result };
    } catch (error) {
      logger.error('Get cell data error:', error);
      throw error;
    }
  });
}

module.exports = { registerDataExplorerHandlers };
